### PR TITLE
DarkSea Implementation Bitvector Abstraction

### PR DESCRIPTION
**Pull request by cyruliu**
Hi @danieldietsch,

I tried to squash the commits on my local, and the PR lost git diff tracks on github (maybe I used the wrong method), it was forced to close after I pushed to its remote, so I recovered it back and create this second one, sorry about the noise. I didn't know there will be a `squash and merge` option for PR here on Github, it should appear after we resolve the conflicts in the end, should we try to resolve the conflict files first?

Best,
Cyrus

### CODE REVIEW

File: trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/FlatSymbolTable.java [hunk 1]
```diff
@@ -419,6 +419,8 @@ public String applyMultiparseRenaming(final String filePath, final String origin
 	public String getCIdForBoogieId(final String boogieId) {
 		return mBoogieIdToCId.get(boogieId);
 	}
+
```
Comment (danieldietsch) [inline trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/FlatSymbolTable.java]: Can you remove additional whitespace?

File: trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/FlatSymbolTable.java [hunk 2]
```diff
@@ -470,6 +472,14 @@ public void addBoogieCIdPair(final String boogieIdentifier, final String cIdenti
 		return Collections.unmodifiableMap(mBoogieIdToCId);
 	}
 
+/**
+ * @cyrus, to get the declared global variables.
+ * @return global scope.
```
Comment (danieldietsch) [inline trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/FlatSymbolTable.java]: Please do not use your name in a method, it just confuses people that do not know you. Also, you should not expose a writable variant of the global scope. Wrap it in Collections.unmodifiableMap()

File: trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/CHandler.java [hunk 3]
```diff
@@ -728,18 +743,29 @@ public Result visit(final IDispatcher main, final IASTBinaryExpression node) {
 		final ILocation loc = mLocationFactory.createCLocation(node);
 		final ExpressionResult leftOperand = (ExpressionResult) main.dispatch(node.getOperand1());
 		final ExpressionResult rightOperand = (ExpressionResult) main.dispatch(node.getOperand2());
-
+		//@CL, with binary expression, we check bitwise operator first
```
Comment (danieldietsch) [inline trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/CHandler.java]: What does @CL mean?

File: trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/CHandler.java [hunk 4]
```diff
@@ -728,18 +743,29 @@ public Result visit(final IDispatcher main, final IASTBinaryExpression node) {
...
-			final CType lType = leftOperand.getLrValue().getCType().getUnderlyingType();
-			final ExpressionResult rightOperandSwitched = mExprResultTransformer
-					.makeRepresentationReadyForConversionAndRexBoolToInt(rightOperand, loc, lType, node);
-			builder.addAllIncludingLrValue(rightOperandSwitched);
-			return makeAssignment(loc, leftOperand.getLrValue(), leftOperand.getNeighbourUnionFields(), builder.build(),
+			
+			//@Cyrus, debug. In this case we only transform with and-rule: r= a&b => r<=b, r<=a, they are positive, 
+			// and rhs is a bitwise binary expression.
+			
+			mLogger.debug("left hand side expression in assignment: "+ leftOperand.getLrValue().toString());
```
Comment (danieldietsch) [inline trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/CHandler.java]: remove this debug log message

File: trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/CHandler.java [hunk 5]
```diff
@@ -750,9 +776,17 @@ public Result visit(final IDispatcher main, final IASTBinaryExpression node) {
...
-			final ExpressionResult rl = mExprResultTransformer.switchToRValue(leftOperand, loc, node);
-			final ExpressionResult rr = mExprResultTransformer.switchToRValue(rightOperand, loc, node);
-			return mCExpressionTranslator.handleRelationalOperators(loc, node.getOperator(), rl, rr);
+		/*
+		 * if (isBit & (node.getOperand1() instanceof IASTIdExpression)) {
+		 * System.out.println("---get bitwise in realational operators: "+node.toString(
+		 * )); return BitabsTranslation.abstractRelational(this, mProcedureManager,
+		 * mDeclarations, mExpressionTranslation, mNameHandler, mAuxVarInfoBuilder,
+		 * mSymbolTable, mExprResultTransformer, main, mLocationFactory, node); } else {
+		 */
```
Comment (danieldietsch) [inline trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/CHandler.java]: remove commented-out code

File: trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/CHandler.java [hunk 6]
```diff
@@ -1923,11 +1957,11 @@ private Result handleStringLiteralExpression(final ILocation loc, final IDispatc
 			ultimateAllocCall = pair.getSecond();
 
 		} else {
-			auxvar = mAuxVarInfoBuilder.constructGlobalAuxVarInfo(actualLoc, pointerType, SFO.AUXVAR.STRINGLITERAL);
+		auxvar = mAuxVarInfoBuilder.constructGlobalAuxVarInfo(actualLoc, pointerType, SFO.AUXVAR.STRINGLITERAL);
 			addressRValue = new RValueForArrays(auxvar.getExp(), arrayType);
 			// the declaration of the variable that corresponds to a string literal has to
 			// be made global
-			mStaticObjectsHandler.addGlobalVarDeclarationWithoutCDeclaration(auxvar.getVarDec());
+		mStaticObjectsHandler.addGlobalVarDeclarationWithoutCDeclaration(auxvar.getVarDec());
```
Comment (danieldietsch) [inline trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/CHandler.java]: fix indentation

File: trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/expressiontranslation/BitabsTranslation.java [hunk 7]
```diff
@@ -0,0 +1,1042 @@
+/*
+ * Copyright (C) 2015 Matthias Heizmann (heizmann@informatik.uni-freiburg.de)
+ * Copyright (C) 2015 Thomas Lang
+ * Copyright (C) 2015 University of Freiburg
+ *
+ * This file is part of the ULTIMATE CACSL2BoogieTranslator plug-in.
```
Comment (danieldietsch) [inline trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/expressiontranslation/BitabsTranslation.java]: add yourself to copyright header, extend times of others if you did not change substantial parts

File: trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/expressiontranslation/BitabsTranslation.java [hunk 8]
```diff
@@ -0,0 +1,1042 @@
...
+		varCounter = 0;
+	}
+
+	public Expression abstractAnd(final ILocation loc, final int op, final Expression left, final CPrimitive typeLeft,
+			final Expression right, final CPrimitive typeRight, final IASTNode hook) {
+		final String funcname = "bitwiseAnd";
+		// if decides_to_apply(CYRUS_AND_0_LEFT, left, right)
+		// If left is equal literal 0 or right is equal literal 0.
+		Expression literal_0 = new IntegerLiteral(loc, BoogieType.TYPE_INT, "0");
+		Expression literal_1 = new IntegerLiteral(loc, BoogieType.TYPE_INT, "1");
```
Comment (danieldietsch) [inline trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/expressiontranslation/BitabsTranslation.java]: use Ultimate coding convention

**Review by danieldietsch — CHANGES_REQUESTED** [review]
Lets wait for the results of the nightly build, then we should be able to merge.

**Comment by cyruliu** [comment]
> Lets wait for the results of the nightly build, then we should be able to merge.

Thanks, will work on the changes.

**Comment by danieldietsch** [comment]
Actually, I had a closer look, it will take some more changes to make this mergeable. Ill look into it later.

**Review by danieldietsch — CHANGES_REQUESTED** [review]
Can you rename all the local variables in `BitabsTranslation` to fit Ultimate's coding convention? Local vars should be lower camel case without underscores, etc.

File: trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/expressiontranslation/BitabsTranslation.java (hunk 8 above)
Comment (cyruliu) [reply trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/expressiontranslation/BitabsTranslation.java]: Working on it.

File: trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/expressiontranslation/BitabsTranslation.java [hunk 9]
```diff
@@ -0,0 +1,884 @@
...
+
+		}
+		return builder.build();
+	}
+
+	public Result abstractRelational(final CHandler chandler, final ProcedureManager mProcedureManager,
+			final ArrayList<Declaration> mDeclarations, final ExpressionTranslation mExpressionTranslation,
+			final INameHandler mNameHandler, final AuxVarInfoBuilder mAuxVarInfoBuilder,
+			final FlatSymbolTable symbolTable, final ExpressionResultTransformer exprResultTransformer,
+			final IDispatcher main, final LocationFactory locationFactory, final IASTBinaryExpression node) {
```
Comment (danieldietsch) [inline trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/expressiontranslation/BitabsTranslation.java]: the `m` prefix is only for global members.

**Review by danieldietsch — CHANGES_REQUESTED** [review]
Looks good, only one small thing left. 
Refactoring the whole thing s.t. it uses existing functions is a separate task :)

**Review by danieldietsch — APPROVED** [review]

**Pull request merged**